### ISSUE TITLE
Method of Applying Updates is Incorrect

### ISSUE THREAD

**Issue by kulvirs** [Dabblr/Concurrent-Document-Editor#5]
#### Client
Currently, when a client sends a list of changes, we are assuming that they are already transformed relative to each other. For example, if a client wishes to add the string "ab" at the start of the file, their list of changes will look like this:
```
{
	"type":"insert",
	"position":0,
	"value":"a"
},
{
	"type":"insert",
	"position":1,
	"value":"b"
}
```
Notice how the insertion of 'b' is already offset to position 1 to account for the insertion of 'a' at position 0. 

#### Server
When the server receives a list of changes for a certain revision number from the client, it transforms each change in the client's list on top of all the changes that have been made to the file since the given revision number. However, since the changes in the client's list are already transformed relative to each other, this can sometimes result in an incorrect final transformation. The examples below illustrate some situations where this occurs.

### Example 1
Suppose the latest file content at Revision i was 'y'.
Client 1 decides to insert 'x' and delete 'y' by sending the following changes:
```
{
	"type":"insert",
	"position":0,
	"value":"x"
},
{
	"type":"delete",
	"position":1,
	"value":"y"
}
```
Since there have been no other changes to the file since Revision i the changes are applied without transformation and the file content at the latest Revision, i+1, is now 'x'.  

Now suppose Client 2 also wishes to post updates on top of Revision i, and has not yet pulled the changes of Revision i+1. This means they still believe the file content to be 'y'. Client 2 wishes to insert the string 'ab' at the start, so that the new file content is 'aby'. They send the following changes:
```
{
	"type":"insert",
	"position":0,
	"value":"a"
},
{
	"type":"insert",
	"position":1,
	"value":"b"
}
```
The server receives these changes and has to transform them on top of the previous changes that have been made since Revision i, which are: `(insert, 0, 'x')` and `(delete, 1, 'y')`.  

This means `(insert, 0, 'a')` gets transformed to `(insert, 1,  'a')`  on top of `(insert, 0, 'x')`  and remains the same on top of `(delete, 1, 'y')`. This updates the file content to 'xa'.  

`(insert, 1, 'b')` gets transformed to `(insert, 2, 'b')` on top of `(insert, 0, 'x')`  but then gets transformed again to `(insert, 1, 'b')`  on top of  `(delete, 1, 'y')`. This updates the file content to 'xba', which does not preserve the original intention of having 'b' come after 'a'. 

### Example 2
When two clients both delete at the same position, the operational transformation algorithm dictates that only one delete gets applied. However, our current system of applying a list of updates results in incorrect values when this happens.

Suppose the file content at Revision i is 'abc'.
Client 1 decides to delete 'b' by sending the following changes:
```
{
	"type":"delete",
	"position":1,
	"value":"b"
}
```
After this change is applied, the file content at Revision i+1 is now 'ac'.

Suppose Client 2 also is working on Revision i and decides to delete 'b' and append 'd' to the end of the string. They send the following changes:
```
{
	"type":"delete",
	"position":1,
	"value":"b"
},
{
	"type":"insert",
	"position":2,
	"value":"d"
}
```
Note how the insertion is already transformed to position 2 to offset for the deletion at position 1.

The server receives these changes and has to transform them on top of the previous changes since Revision i which is just `(delete, 1, 'b')`. 

The client's `(delete, 1, 'b')` gets removed because it is a duplicate of the previous deletion.
However, the `(insert, 2, 'd')` gets transformed on top of the previous change `(delete, 1, 'b')`  to become `(insert, 1, 'd')`. The new file content becomes 'adc' which is not what the client intended, 'd' should be at the end of the string.

### Proposed Solution
The way I see it, these problems could easily be resolved if the changes sent by client are not transformed relative to each other. So if a client wishes to insert 'ab' at the start of the string, they would send the following list to the server:
```
{
	"type":"insert",
	"position":0,
	"value":"a"
},
{
	"type":"insert",
	"position":0,
	"value":"b"
}
```
After the server transforms a changes on top of the existing ones, it adds the newly transformed change to the list of existing changes. This means subsequent changes in the client's list will get transformed on top of the previous changes and then on top of the changes that came before it in its own list. I think this will preserve the proper ordering of transformations better.

**Comment by okjuan** [comment]
I'm gonna write some ideas down here while I have them.

#### Example 1
Upon further reading about OT, I learned two important things *[1]* *[2]*:
1. Operations are computed by **calculating a diff** between document revisions, rather than by tracking keyboard events
2. Operations can (and it looks like they should) describe **all characters** in a document revision, not just the insertions and deletions!
    * If we abide by this, we might have to implement the `retain` operation, which indicates that `x` characters from the parent revision be copied (retained) in the new revision; however, I suspect that specifying the position of insertions and deletions serves the same purpose. If this is true, then we could get away without using `retain`.

This means that, with respect to the Example 1 above, a revision that changes a document from contents `"y"` to `"x"` would represent the changes with a list of operations `[insert("x", 0)]`, instead of `[insert("x", 0), delete("y")]`. This illustrates point 2.
* The `delete("y")` becomes redundant. The list of changes already comprehensively describes the contents of the document, thus, the character `"y"` does not show up in the new revision because the list of changes did not retain it (or insert a new `"y"`).
* Notice that, if we decide to use `retain` operations, the position (`0`) in the insertion is actually also redundant, since the character position would be implicit in the sum of the operation position in the list and the preceding `retain` values.

*Side-note*: In this example, it might seem like our new way of using operations to describe document revisions is terser, but this isn't really the case.

#### Example 2
Example 2 shows a different problem, but, luckily, I think we can solve it by updating our OT functions to return two values (instead of one) and using that second value in subsequent transformations. More specifically, we want `OT(a,b)` (where `a` is the client change and `b` is the server change) to return `(a', b')`, instead of just `a'` (where `a'` is the transformed client change and `b'` is the transformed server change). See the *Transformation* section of *[2]* for an in-depth explanation on the reasoning behind this; in a nutshell, OT only resolves divergence of one client change and one server change, so, when resolving multiple changes from the client and/or server, we have to compose OT ops with results of previous OT ops to properly bring client and server back together. If this was just confusing, I think the idea is made clearer by seeing how it applies to the Example 2.

In Example 2, an issue arises when both the client and server delete the same character. In short, the way we are currently using OT results in an unexpected state (see the description of Example 2 above for details). To solve this, intuitively, we want to apply the duplicate deletion only once and **avoid transforming subsequent client changes onto it**, since they are already defined in context of that deletion.

More formally, in the diagrams below, let the client changes be `a` and `b` and the server change be `c`. Our goal is to find `a'` and `b'` so that we can apply them to `c`, effectively incorporating all client changes to the server. 
* We can get `a'` with `OT(a,b)`.
  * Our current implementation calls the right to OT, but receives `a' = a`, which turns out to be wrong in this case.
* We can get `b'` with `OT(b, c')`.
  * However, our current implementation gets `b'` from `OT(b, c)`. In many cases, this turns out to be fine because `c = c'`, but not in this case!


<img 
    style="display: inline-block;"
    width="280"  
    src="https://user-images.githubusercontent.com/13096220/49704075-20114f80-fbc3-11e8-94b0-57f8df2f986b.png" 
/>
<img  
    style="display: inline-block;"
    width="285"  
    src="https://user-images.githubusercontent.com/13096220/49704078-230c4000-fbc3-11e8-8477-99bd91f2a64f.png"
/>
<img  
    style="display: inline-block;"
    width="289"  
    src="https://user-images.githubusercontent.com/13096220/49704079-256e9a00-fbc3-11e8-8a31-a76d3921c739.png"
/>
<img  
    style="display: inline-block;"
    width="308"  
    src="https://user-images.githubusercontent.com/13096220/49704080-269fc700-fbc3-11e8-87d5-014d9aac918e.png"
/>

The green, dotted line is the transformed server operation `c'`, which we get from `OT(a,c)`.

#### Questions
##### Duplicate Insertions
*What if the client and server both make the same insertion, in the same way that they make the same deletion in Example 2? Should we make sure that the insertion is applied only once?*

No, we should apply both insertions because we don't want to treat duplicate insertions as identical (though we do want to treat duplicate deletions as identical). The motivation for this should be clear from the UX perspective of a shared document editor; if the insertions really are redundant duplicates, we'd leave it to the users to resolve manually at their discretion.

##### Duplicate Deletions
*What if duplicate deletions happen at different places in the client and server change history? Example 2 shows the simplest case in which we have duplicate deletions and how our current use of OT fails to handle it. We shouldn't assume that fixing our use of OT to handle Example 2 would handle all cases where the same deletion is performed by both client and server.*
 
Idk the answer to this yet. However, it's important to keep in mind that client and server changes will be ordered by their position of occurrence in the document -- that is, an operation on a character at position `i` will always precede some other operation on a character at a position `j` if `i < j`.

*UPDATE*: The following cases work nicely:
Case 1: 
Initial state: `"abc"`

```
// Client changes:
[ { "type":"delete",   "position":1,   "value":"b"} ]

// Server changes:
[
  { "type":"insert",   "position":3,   "value":"d" },
  { "type":"delete",   "position":0,   "value":"b" }
]
```
Resulting state: `"acd"`.

Case 2
Initial state: `"abc"`
```
// Client changes:
[ { "type":"delete",   "position":1,   "value":"b"} ]

// Server changes:
[
  { "type":"delete",   "position":0,   "value":"a" },
  { "type":"delete",   "position":0,   "value":"b" }
]
```
Resulting state: `"c"`.

#### References
* *[1]*: [Operational Transformation: Operations](http://fitzgeraldnick.com/2011/04/05/operational-transformation-operations.html)
* *[2]*: [Understanding and Applying Operational Transformation](http://www.codecommit.com/blog/java/understanding-and-applying-operational-transformation)
